fdbclient/NativeAPI.actor.cpp

```diff
@@ -2375,15 +2375,44 @@ Reference<WatchMetadata> DatabaseContext::getWatchMetadata(int64_t tenantId, Key
 }
 
 void DatabaseContext::setWatchMetadata(Reference<WatchMetadata> metadata) {
-	watchMap[std::make_pair(metadata->parameters->tenant.tenantId, metadata->parameters->key)] = metadata;
+	const WatchMapKey key(metadata->parameters->tenant.tenantId, metadata->parameters->key);
+	watchMap[key] = metadata;
+	// NOTE Here we do *NOT* update/reset the reference count for the key, see the source code in getWatchFuture
+}
+
+int32_t DatabaseContext::increaseWatchRefCount(const int64_t tenantID, KeyRef key) {
+	const WatchMapKey mapKey(tenantID, key);
+	if (watchCounterMap.count(mapKey) == 0) {
+		watchCounterMap[mapKey] = 0;
+	}
+	const auto count = ++watchCounterMap[mapKey];
+	return count;
+}
+
+int32_t DatabaseContext::decreaseWatchRefCount(const int64_t tenantID, KeyRef key) {
+	const WatchMapKey mapKey(tenantID, key);
+	if (watchCounterMap.count(mapKey) == 0) {
+		// Key does not exist. The metadata might be removed by deleteWatchMetadata already.
+		return 0;
+	}
+	const auto count = --watchCounterMap[mapKey];
+	ASSERT(watchCounterMap[mapKey] >= 0);
+	if (watchCounterMap[mapKey] == 0) {
+		getWatchMetadata(tenantID, key)->watchFutureSS.cancel();
```

Comment:
Do we need to cancel the future here? Just causing the metadata to be deleted should be sufficient.

Reply:
I am unclear why, but it seems that manually canceling the actor is a must in order to make corresponding DatabaseContext destruct properly.

Reply:
I think one reason is that

watchFutureSS = wait(watchStorageServerResp)

watchStorageServerResp contains a reference to the metadata.

Thus, if we do not manually cancel watchStorageServerResp, it itself will hold a reference to the metadata, the reference count of the metadata will not be zero.